src/pvcam.cpp

```diff
@@ -0,0 +1,1200 @@
+            char msg[ERROR_MSG_LEN];                                           \
+            pl_error_message(code, msg);                                       \
+            logger(msg);                                                       \
+            action;                                                            \
+        }                                                                      \
+    } while (0)
+
+#define PVCAM(e, mtx)                                                          \
+    PVCAM_INNER(                                                               \
+      e, mtx, LOGE, throw std::runtime_error("PVCAM API call failed."))
+#define PVWARN(e, mtx) PVCAM_INNER(e, mtx, LOG, )
+
+namespace {
+/// Utilities
+
+// Maps Acquire SampleType to GenICam pixel format strings.
+const std::unordered_map<SampleType, PL_IMAGE_FORMATS>
+  sample_type_to_pixel_format{
+      { SampleType_u8, PL_IMAGE_FORMAT_MONO8 },
+      { SampleType_u16, PL_IMAGE_FORMAT_MONO16 },
+  };
+
+const std::unordered_map<PL_IMAGE_FORMATS, SampleType>
+  pixel_format_to_sample_type{
+      { PL_IMAGE_FORMAT_MONO8, SampleType_u8 },
+      { PL_IMAGE_FORMAT_MONO16, SampleType_u16 },
+  };
+
+/**
+ * @brief Checks that the given camera id is within the range of a signed
+ * 16-bit.
+ * @param id The camera id to check.
+ * @throws std::runtime_error if the id is out of range.
+ */
+void
+check_pvcam_camera_id(uint64_t id)
+{
+    constexpr uint64_t limit = std::numeric_limits<int16_t>::max();
+    EXPECT(id <= limit, "Expected a value at most %llu. Got %llu.", limit, id);
+}
+
+void
+camera_properties_to_rgn_type(CameraProperties* properties, rgn_type& roi)
+{
+    auto binning = (uint16_t)properties->binning;
+    roi = {
+        .s1 = (uint16_t)properties->offset.x,
+        .s2 = (uint16_t)(properties->offset.x + properties->shape.x * binning - 1),
+        .sbin = binning,
+        .p1 = (uint16_t)properties->offset.y,
+        .p2 = (uint16_t)(properties->offset.y + properties->shape.y * binning - 1),
+        .pbin = binning,
+    };
+}
+
+void
+rgn_type_to_camera_properties(rgn_type& roi, CameraProperties* properties)
+{
+    properties->binning = (uint8_t)roi.pbin;
+
+    properties->offset = {
+        .x = roi.s1,
+        .y = roi.p1,
+    };
+    properties->shape = {
+        .x = (uint32_t)((roi.s2 - roi.s1 + 1) / properties->binning),
+        .y = (uint32_t)((roi.p2 - roi.p1 + 1) / properties->binning),
+    };
+}
+
+// forward
+struct PVCamCamera;
+
+struct CallbackContext
+{
+    int16_t hcam;
+
+    std::shared_ptr<std::mutex> api_mutex;
+    std::shared_ptr<std::mutex> frame_mutex;
+    std::condition_variable cv;
+    std::atomic<int> frames_available;
+
+    uint8_t** frame;
+    FRAME_INFO info;
+};
+
+void
+callback_handler(FRAME_INFO* frame_info, void* context)
+{
+    auto* ctx = (CallbackContext*)context;
+    std::unique_lock<std::mutex> frame_lock(*ctx->frame_mutex);
```

Comment:
Nit: since you don't do any manual locking or unlocking, `scoped_lock` is enough here and simpler than `unique_lock`, so I have a mild preference for it.